samcli/lib/sync/flows/rest_api_sync_flow.py

```diff
@@ -48,15 +51,156 @@ def __init__(
+        self._delete_deployments(prev_dep_ids)
+
+    def _update_api(self) -> None:
+        """
+        Update the API content
+        """
+        LOG.debug("%sTrying to update RestAPI through client", self.log_prefix)
+        response_put = cast(
+            Dict,
+            self._api_client.put_rest_api(restApiId=self._api_physical_id, mode="overwrite", body=self._swagger_body),
+        )
+        LOG.debug("%sPut RestApi Result: %s", self.log_prefix, response_put)
+
+    def _create_deployment(self) -> Optional[str]:
+        """
+        Create a deployment using the updated API and record the created deployment ID
+
+        Returns
+        ----------
+        Optional[str]: The newly created deployment ID
+        """
+        LOG.debug("%sTrying to create a deployment through client", self.log_prefix)
+        response_dep = cast(
+            Dict, self._api_client.create_deployment(restApiId=self._api_physical_id, description="Created by SAM Sync")
+        )
+        new_dep_id = response_dep.get("id")
+        LOG.debug("%sCreate Deployment Result: %s", self.log_prefix, response_dep)
+        return new_dep_id
+
+    def _collect_stages(self) -> Set[str]:
+        """
+        Collect all stages needed to be updated
+
+        Returns
+        ----------
+        Set[str]: The set of stage names to be updated
+        """
+        # Get the stage name associated with the previous deployment and update stage
+        # Stage needs to be flushed so that new changes will be visible immediately
+        api_resource = get_resource_by_id(self._stacks, ResourceIdentifier(self._api_identifier))
+        stage_resources = get_resource_ids_by_type(self._stacks, "AWS::ApiGateway::Stage")
+
+        stages = set()
+        # If it is a SAM resource, get the StageName property
+        if api_resource:
+            if api_resource.get("Type") == "AWS::Serverless::Api":
+                # The customer defined stage name
+                stage_name = api_resource.get("Properties").get("StageName")  # type: ignore
+                stages.add(cast(str, stage_name))
+
+                # The Stage stage
+                if stage_name != "Stage":
+                    response_sta = cast(Dict, self._api_client.get_stages(restApiId=self._api_physical_id))
+                    for item in response_sta.get("item"):  # type: ignore
+                        if item.get("stageName") == "Stage":
+                            stages.add("Stage")
+
+        # For both SAM and ApiGateway resource, check if any refs from stage resources
+        for stage_resource in stage_resources:
+            # RestApiId is a required field in stage
+            stage_dict = get_resource_by_id(self._stacks, stage_resource)
+            if stage_dict:
+                rest_api_id = stage_dict.get("Properties").get("RestApiId")  # type: ignore
+                dep_id = stage_dict.get("Properties").get("DeploymentId")  # type: ignore
```

Comment:
Same comment as above, I am not sure if these properties are required. So let's make sure they are not None before working on them.

Reply:
They are also required properties, I don't think we need to do the job that sam validate does here. These two values being None doesn't hurt the sync process so we don't need to have any handling here. The only case that this code crashes would be when a customer have a stage resource that doesn't have the Properties field. Again I think this validation would be too unnecessary. Let me know your thoughts!

Reply:
Yeah I don't think we need to run validation for `Properties`. But for the sake of safetiness, we can use `stage_dict.get("Properties", {}).get("DeploymentId")` so that even if they have an empty `Properties` this won't end up with any exception

Reply:
Good point! I'll add that